apps/heft/src/pluginFramework/HeftEventPluginBase.ts

```diff
@@ -0,0 +1,173 @@
+        await handleEventActionsAsync(
+          HeftEvent.clean,
+          clean.properties,
+          this.handleCleanEventActionsAsync.bind(this)
+        );
+      });
+    });
+
+    heftSession.hooks.build.tap(this.pluginName, (build: IBuildStageContext) => {
+      build.hooks.preCompile.tap(this.pluginName, (preCompile: IPreCompileSubstage) => {
+        preCompile.hooks.run.tapPromise(heftStageTap, async () => {
+          await handleEventActionsAsync(
+            HeftEvent.preCompile,
+            build.properties,
+            this.handleBuildEventActionsAsync.bind(this)
+          );
+        });
+      });
+
+      build.hooks.compile.tap(this.pluginName, (compile: ICompileSubstage) => {
+        compile.hooks.run.tapPromise(heftStageTap, async () => {
+          await handleEventActionsAsync(
+            HeftEvent.compile,
+            build.properties,
+            this.handleBuildEventActionsAsync.bind(this)
+          );
+        });
+      });
+
+      build.hooks.bundle.tap(this.pluginName, (bundle: IBundleSubstage) => {
+        bundle.hooks.run.tapPromise(heftStageTap, async () => {
+          await handleEventActionsAsync(
+            HeftEvent.bundle,
+            build.properties,
+            this.handleBuildEventActionsAsync.bind(this)
+          );
+        });
+      });
+
+      build.hooks.postBuild.tap(this.pluginName, (postBuild: IPostBuildSubstage) => {
+        postBuild.hooks.run.tapPromise(heftStageTap, async () => {
+          await handleEventActionsAsync(
+            HeftEvent.postBuild,
+            build.properties,
+            this.handleBuildEventActionsAsync.bind(this)
+          );
+        });
+      });
+    });
+
+    heftSession.hooks.test.tap(this.pluginName, (test: ITestStageContext) => {
+      test.hooks.run.tapPromise(heftStageTap, async () => {
+        await handleEventActionsAsync(
+          HeftEvent.test,
+          test.properties,
+          this.handleTestEventActionsAsync.bind(this)
+        );
+      });
+    });
+  }
+
+  protected handleCleanEventActionsAsync(
+    heftEvent: HeftEvent,
+    heftEventActions: THeftEventAction[],
+    logger: ScopedLogger,
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    properties: ICleanStageProperties
+  ): Promise<void> {
+    return Promise.resolve();
+  }
+
+  protected handleBuildEventActionsAsync(
+    heftEvent: HeftEvent,
+    heftEventActions: THeftEventAction[],
+    logger: ScopedLogger,
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    properties: IBuildStageProperties
+  ): Promise<void> {
+    return Promise.resolve();
+  }
+
+  protected handleTestEventActionsAsync(
+    heftEvent: HeftEvent,
+    heftEventActions: THeftEventAction[],
+    logger: ScopedLogger,
+    heftSession: HeftSession,
+    heftConfiguration: HeftConfiguration,
+    properties: ITestStageProperties
+  ): Promise<void> {
+    return Promise.resolve();
+  }
```

Comment:
Seems like these should just be abstract and required to be implemented in the child class.

Reply:
Yeah I had thought about that, went this route to provide a default no-op so that plug-ins that don't support other `HeftEvent` values don't need to have implementations